### PR TITLE
[hail][bugfix] Fix bug in correctness of n_discordant

### LINKED ISSUE #7632: Concordance function may give wrong n_discordant values?
Hi, 

I'm using the concordance function to compare two sets of data, and I feel the n_discordant (last column) is not correct. 

For example: 
```
chr1:930314     ["C","T"]       {"locus":{"contig":"chr1","position":930314},"alleles":["C","T"]}       {"locus":{"contig":"chr1","position":930314},"alleles":["C","T"]}       [[0,0,0,0,0],[0,0,21,1,0],[0,0,2057,0,0],[0,0,0,91,0],[0,0,0,0,3]]      2172
chr1:946538     ["G","A"]       {"locus":{"contig":"chr1","position":946538},"alleles":["G","A"]}       {"locus":{"contig":"chr1","position":946538},"alleles":["G","A"]}       [[0,0,0,0,0],[0,0,5,3,0],[0,0,1868,1,1],[0,0,0,279,0],[0,0,0,0,16]]     2170
chr1:946653     ["G","A"]       {"locus":{"contig":"chr1","position":946653},"alleles":["G","A"]}       {"locus":{"contig":"chr1","position":946653},"alleles":["G","A"]}       [[0,0,0,0,0],[0,0,856,275,66],[0,0,386,74,7],[0,0,16,415,33],[0,0,0,3,42]]      1898
```

In the first example, I thought the n_discordant should be 0 if the `concordance` field is correct, isn't it?

The code I was using: 
`global_GA_both, samples_GA_both, SNPs_GA_both = hl.concordance(mt_exome, mt_GAsP_ft)`

The Hail version:
```Running on Apache Spark version 2.4.3
SparkUI available at http://spark-master:4040
Welcome to
     __  __     <>__
    / /_/ /__  __/ /
   / __  / _ `/ / /
  /_/ /_/\_,_/_/_/   version 0.2.26-2dcc3d963867
LOGGING: writing to Concordance_2019_11_28_hail.log
```

When I was using google Terra Hail 0.2.11-daed180b84d8, I didn't have this issue. The output didn't have `left_row` or `right_row`.

Cheers,
Qinqin

### PR DESCRIPTION
**Pull request by tpoterba**
fixes #7632

### CODE REVIEW

File: hail/python/test/hail/methods/test_qc.py [hunk 1]
```diff
@@ -135,6 +135,11 @@ def test_concordance(self):
         cols_conc.write('/tmp/foo.kt', overwrite=True)
         rows_conc.write('/tmp/foo.kt', overwrite=True)
 
+    def test_concordance_n_discordant(self):
+        dataset = get_dataset()
+        _, cols_conc, rows_conc = hl.concordance(dataset, dataset)
+        assert cols_conc.aggregate(hl.agg.count_where(cols_conc.n_discordant != 0)) == 0
```
Comment (akotlar) [inline hail/python/test/hail/methods/test_qc.py]: Should we test the complementary case ( > 0 )?

**Comment by tpoterba** [comment]
yes, good idea. I added a hard-coded test of both the arrays and n_discordant.

File: hail/python/hail/methods/qc.py [hunk 2]
```diff
@@ -442,10 +442,16 @@ def get_idx(struct):
...
+    for i in range(5):
+        for j in range(5):
+            if i > 1 and j > 1 and i != j:
+                discordant_indices.add(i + 5 * j)
+
     def n_discordant(counter):
         return hl.sum(
             hl.array(counter)
-                .filter(lambda tup: ~hl.literal({i ** 2 for i in range(5)}).contains(tup[0]))
+                .filter(lambda tup: hl.literal(discordant_indices).contains(tup[0]))
```
Comment (johnc1231) [inline hail/python/hail/methods/qc.py]: Just to make sure I understand, the issue here was that the old method didn't handle 0 and 1 correctly? Hence the need to do the `if i > 1 and j > 1` big above?

Comment (tpoterba) [reply hail/python/hail/methods/qc.py]: no, it was just totally wrong!

Comment (tpoterba) [reply hail/python/hail/methods/qc.py]: but I think we intended to make n_discordant exclude missingness/no data

**Pull request merged**